### PR TITLE
Customization of file watchers

### PR DESCRIPTION
**Pull request by kreuli**
This PR covers improvement #153: https://github.com/SpartanJ/efsw/issues/153

It adds a concept of 'WatcherOptions' that can be used to customize file watchers. Two such options have been implemented, to customize the Windows implementation of FileWatcher: an option to specify the buffer size for the ReadDirectoryChangesW API and another option to reduce the events specified in NotificationFilter. This can be used to adjust the FileWatchers for Windows, in case the currently used buffer of 63kB is not enough and the FileWatcher therefore does not work as of now.

Additionally, the return code of ReadDirectoryChangesW is read and LastError is set to a newly introduced error. This can be used to detect issues with defining a too small buffer size.

### CODE REVIEW

**Comment by SpartanJ** [comment]
Thanks for your contribution! I'll leave some minor comments on the source changes.

**Comment by SpartanJ** [comment]
@kreuli are you interested in implementing some of the changes requested? Let me know so I know how to proceed. Thank you.

**Comment by kreuli** [comment]
Hello Martin,yes, I‘d like to do that. Should I already see comments with requests for changes somewhere, because I am not aware of any.Cheers,UlrichAm 04.07.2023 um 15:56 schrieb Martín Lucas Golini ***@***.***>:﻿
@kreuli are you interested in implementing some of the changes requested? Let me know so I know how to proceed. Thank you.

**Comment by SpartanJ** [comment]
Hi @kreuli !
You can see the comments above your comment in the Pull Request. I guess you answered from your email: https://github.com/SpartanJ/efsw/pull/162
Also in the file changes there is more visibility: https://github.com/SpartanJ/efsw/pull/162/files
You can also subscribe to the PR so you can get notifications each time I write here.
Thanks

**Comment by kreuli** [comment]
Your review is shown as pending, which is why I can‘t see your code comments, I guess.

File: README.md [hunk 1]
```diff
@@ -78,6 +78,10 @@ efsw::WatchID watchID = fileWatcher->addWatch( "/tmp", listener, true );
 // Adds another directory to watch. This time as non-recursive.
 efsw::WatchID watchID2 = fileWatcher->addWatch( "/usr", listener, false );
 
+// For Windows, adds another watch, specifying to use a bigger buffer, to not miss events
+// (do not use for network locations, see efsw.hpp for detaisl!).
```
Comment (solarispika) [inline README.md]: A typo here: `detaisl` should be `details`.

**Review by SpartanJ — CHANGES_REQUESTED** [review]

**Comment by SpartanJ** [comment]
@kreuli sorry, I forgot about this. I just published the review. Thank you

File: include/efsw/efsw.h [hunk 2]
```diff
@@ -103,11 +103,13 @@ efsw_watcher EFSW_API efsw_create(int generic_mode);
...
 void EFSW_API efsw_release(efsw_watcher watcher);
 
-/// Retreive last error occured by file-watcher
+/// Retrieve last error occured by file-watcher
 EFSW_API const char* efsw_getlasterror();
 
-/// Add a directory watch. Same as the other addWatch, but doesn't have recursive option.
-/// For backwards compatibility.
+/// Reset file-watcher last error
+EFSW_API void efsw_clearlasterror();
```
[the remark this answers is not in the document]
Comment (kreuli) [reply include/efsw/efsw.h:130]: Extended the C interface accordingly. Also added a new Standard C test application, which is basically the C++ test application but in C.

File: include/efsw/efsw.hpp [hunk 3]
```diff
@@ -90,27 +92,52 @@ typedef Actions::Action Action;
...
 class EFSW_API Log {
   public:
 	/// @return The last error logged
 	static std::string getLastErrorLog();
 
+	/// @return The code of the last error logged
+	static Error getLastErrorCode();
+
+	/// Reset last error
+	static void clearLastError();
```
[the remark this answers is not in the document]
Comment (kreuli) [reply include/efsw/efsw.hpp:117]: I want to be able to check if the "WatcherFailed" error occurecd (the Win32 provided buffer being too small) and to reset the error code.

File: src/efsw/Log.cpp [hunk 4]
```diff
@@ -23,11 +34,15 @@ Error Log::createLastError( Error err, std::string log ) {
...
 			break;
+		case WinReadDirectoryChangesFailed:
+			LastError = "ReadDirectoryChanges failed ( " + log + " )";
+			break;
 		case Unspecified:
 		default:
 			LastError = log;
 	}
 
+	efDEBUG( "%s\n", LastError );
```
[the remark this answers is not in the document]
Comment (kreuli) [reply src/efsw/Log.cpp]: Changed

**Comment by kreuli** [comment]
I've just pushed changes as requested and also have updated the base of my branch to your newest master. There is now also a Standard C test application, basically the same as the C++ test application, to test the C wrapper.

File: include/efsw/efsw.hpp [hunk 5]
```diff
@@ -90,27 +92,54 @@ typedef Actions::Action Action;
...
+	NoError = 0,
 	FileNotFound = -1,
 	FileRepeated = -2,
 	FileOutOfScope = -3,
 	FileNotReadable = -4,
-	FileRemote = -5, /** Directory in remote file system ( create a generic FileWatcher instance to
-						watch this directory ). */
-	Unspecified = -6
+	/// Directory in remote file system
+	/// ( create a generic FileWatcher instance to watch this directory ).
```
Comment (SpartanJ) [inline include/efsw/efsw.hpp:101]: You forgot the first part of the comment: `Directory in remote file system`

File: src/efsw/FileWatcherGeneric.hpp [hunk 6]
```diff
@@ -5,6 +5,8 @@
 #include <efsw/FileWatcherImpl.hpp>
 #include <efsw/WatcherGeneric.hpp>
 #include <vector>
+#include <vector>
+#include <vector>
```
Comment (SpartanJ) [inline src/efsw/FileWatcherGeneric.hpp:9]: Duplicated line.

**Comment by SpartanJ** [comment]
This is looking really good! Thank you so much for this. Please give me a few days since I'm not available to do a proper review right now but meanwhile I'll just send you some 2 minor comments.

**Review by SpartanJ — CHANGES_REQUESTED** [review]

File: src/efsw/FileWatcherGeneric.hpp (hunk 6 above)
Comment (kreuli) [reply src/efsw/FileWatcherGeneric.hpp]: Removed

File: include/efsw/efsw.hpp (hunk 5 above)
Comment (kreuli) [reply include/efsw/efsw.hpp:101]: Did I? I changed the /** style comment for FileRemote to a /// comment, as used in (most of) the other code and added a comment for the new value WatherFailed.

File: src/test/efsw-test.c [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+		char cwd[256];
+		getcwd( cwd, sizeof(cwd) );
+		printf( "CurPath: %s\n", cwd );
+
+		/// starts watching
+		efsw_watch( fileWatcher );
+
+		/// add a watch to the system
+		char path1[256];
+		sprintf(path1, "%s\\test", cwd );
```
Comment (SpartanJ) [inline src/test/efsw-test.c:118]: This is using windows path separator. It will fail on any other OS.

File: src/test/efsw-test.c [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+		handleWatchID( efsw_addwatch_withoptions( fileWatcher, path1, handleFileAction, true, 0, 0, 0 ) );
+
+		/// adds another watch after started watching...
+		sleepMsecs( 100 );
+
+		char path2[256];
+		sprintf(path2, "%s\\test2", cwd );
```
Comment (SpartanJ) [inline src/test/efsw-test.c:125]: Same here.

**Review by SpartanJ — COMMENTED** [review]
This is almost done! I two one minor comment in the new C test implementation.
We are going to need to solve the conflicts, since I had to update the macOS implementation with a new API.
Thank you very much for your contribution!

File: src/test/efsw-test.c (hunk 7 above)
Comment (kreuli) [reply src/test/efsw-test.c]: Changed to use / on non-Windows platforms.

File: src/test/efsw-test.c (hunk 8 above)
Comment (kreuli) [reply src/test/efsw-test.c]: Changed to use / on non-Windows platforms.

**Comment by kreuli** [comment]
Just to be sure: is there something missing from my side?

**Review by SpartanJ — APPROVED** [review]

**Comment by SpartanJ** [comment]
> Just to be sure: is there something missing from my side?

Nothing missing! Thanks for your contribution!

**Pull request merged**